engine/common/synchronization/finalized_snapshot.go

```diff
@@ -0,0 +1,139 @@
+package synchronization
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/lifecycle"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// finalizedSnapshot is a helper structure which contains latest finalized header and participants list
+type finalizedSnapshot struct {
+	head         *flow.Header
+	participants flow.IdentityList
+}
+
+// FinalizedSnapshotCache represents a cached snapshot of the latest finalized header and participants list.
+// It is used in Engine to access latest valid data.
+type FinalizedSnapshotCache struct {
+	mu sync.Mutex
+
+	log                       zerolog.Logger
+	state                     protocol.State
+	identityFilter            flow.IdentityFilter
+	lastFinalizedSnapshot     *finalizedSnapshot
+	finalizationEventNotifier engine.Notifier // notifier for finalization events
+
+	lm                *lifecycle.LifecycleManager
+	shutdownCompleted chan struct{} // used to signal that shutdown of finalization processing loop was completed
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewFinalizedSnapshotCache creates a new finalized snapshot cache.
+func NewFinalizedSnapshotCache(log zerolog.Logger, state protocol.State, participantsFilter flow.IdentityFilter) (*FinalizedSnapshotCache, error) {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
what is the idea behind using a context which never cancels?

Reply:
It is cancelled [here](https://github.com/onflow/flow-go/pull/1068/files#diff-8ae69ca39c72d3cc51859c7a4fa6b620b5cabf9c4b0fc8ac57f9ff1437cfd0c3R107)